akka-contrib/src/main/scala/akka/contrib/pattern/ClusterSingletonManager.scala

```diff
@@ -191,8 +191,9 @@ object ClusterSingletonManager {
 
       def handleInitial(state: CurrentClusterState): Unit = {
         membersByAge = immutable.SortedSet.empty(ageOrdering) ++ state.members.filter(m ⇒
-          m.status == MemberStatus.Up && matchingRole(m))
-        val initial = InitialOldestState(membersByAge.headOption.map(_.address), membersByAge.size)
+          (m.status == MemberStatus.Up || m.status == MemberStatus.Leaving) && matchingRole(m))
+        val safeToBeOldest = !state.members.exists { m ⇒ (m.status == MemberStatus.Down || m.status == MemberStatus.Exiting) }
```

Comment:
That is done on next line, `membersByAge.headOption`.
I guess you are thinking about that is `safeToBeOldest` also when a non-oldest is leaving?
I rather be careful than trying to optimize too much here. All that are leaving are most likely older than current node and therefore potential owners of the singleton instance.